clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -183,22 +183,33 @@ public RecordAppendResult append(TopicPartition tp,
             ByteBuffer buffer = free.allocate(size, maxTimeToBlock);
             synchronized (dq) {
                 // Need to check if producer is closed again after grabbing the dequeue lock.
-                if (closed)
+                if (closed) {
+                    free.deallocate(buffer);
                     throw new IllegalStateException("Cannot send after the producer is closed.");
+                }
+
+                RecordAppendResult appendResult = null;
+                try {
+                    appendResult = tryAppend(timestamp, key, value, callback, dq);
+                } catch (Exception e) {
```

Comment:
Seems like you want `try/finally` instead of `try/catch` here. It may be worth being conservative and having a wider `try/finally` and nulling out the outer variable if we have used the buffer successfully. Something like:

```java
ByteBuffer allocatedBuffer = free.allocate(buffer);
try {
  ...
  // after successfully using the buffer
  allocatedBuffer = null;
} finally {
  if (allocatedBuffer != null)
    free.deallocate(allocatedBuffer)
}